Added code to perform hail size forecasting diagnostic

ywangwof opened the pull request:
**Description**

FV3-HAILCAST hail size forecasting diagnostic (Adams-Selin and Ziegler 2016 MWR). 

Authors: John Henderson and Rebecca Adams-Selin AER, Conrad Ziegler NSSL.

**How Has This Been Tested?**

It was tested on Jet. In order to turn on the HAILCAST package, this namelist block should be added to `input.nml`.

> &fv_diagnostic_nml
>       do_hailcast = .true.
>/

and this entry should be added to `diag_table`

> "gfs_dyn",   "hailcast_dhail_max",    "HAILCAST_DHAIL",    "fv3_history2d", "all", .false., "none", 2

**Checklist:**

Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

lharris4 commented:
This is a very nice addition to FV3 and I am glad to see that has been contributed back to the community. I have a couple of recommendations:

1) I would prefer that the diagnostics code and variables all be contained within the one HAILCAST module. In particular the code being put into fv_arrays.F90 and fv_nggps_diag.F90 should all be put within the one module_diag_hailcast.F90, with only an initialization routine and an execution routine being exposed to the rest of the code base. The files fv_arrays.F90 and fv_nggps_diag.F90 are already too large and unwieldy.

2) Is this implementation of HAILCAST this designed for any particular microphysical scheme? If so this should be explicitly stated and checked for at initialization. Alternately, if the scheme is relatively generic, but needs the microphysics to make available certain fields, this should also be checked for at initialization, with either a warning or a fatal error thrown if the appropriate fields are not found.

Thanks for your contribution to FV3.

bensonr commented:
@ywangwof - @lharris4 requested changes to this PR.  If you are unable or unwilling to address these by the end of January 2022, we will be closing this PR.

ywangwof commented:
Hello @bensonr, we are working on the request. Unfortunately, we are occupied by some time-sensitive projects recently. Could we request for an extension to the end of February for this task? Thanks.

bensonr commented:
@ywangwof - we can leave this open.

lharris4 commented:
Hi, Yunheng. No problem, whenever you are ready.

Lucas

On Tue, Jan 18, 2022 at 4:45 PM Rusty Benson ***@***.***>
wrote:

> @ywangwof <https://github.com/ywangwof> - we can leave this open.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/164#issuecomment-1015862241>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVCM6DCPG6HHM4SC5C3UWXNOPANCNFSM5H57G5YA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lharris4 commented:
This is much better, exactly what I was asking for. As long as this is checked so it compiles and works correctly I would approve this merge. Thanks for working on this.

bensonr commented on model/fv_arrays.F90:
```diff
@@ -61,6 +61,10 @@ module fv_arrays_mod
      real :: efx(max_step), efx_sum, efx_nest(max_step), efx_sum_nest, mtq(max_step), mtq_sum
      integer :: steps
 
+     !for hailcast
+     !integer :: id_hailcast_dhail1, id_hailcast_dhail2, id_hailcast_dhail3, id_hailcast_dhail4, id_hailcast_dhail5
+     !integer :: id_hailcast_diam_mean, id_hailcast_diam_std
+
```
Please remove these commented out lines.

bensonr commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1852 @@
+        integer :: unit, f_unit
+
+        !namelist file for hailcast
+#ifdef INTERNAL_FILE_NML
+        ! Read Main namelist
+        read (input_nml_file,fv_diagnostics_nml,iostat=ios)
+        ierr = check_nml_error(ios,'fv_diagnostics_nml')
+#else
+        f_unit=open_namelist_file()
+        rewind (f_unit)
+        ! Read Main namelist
+        read (f_unit,fv_diagnostics_nml,iostat=ios)
+        ierr = check_nml_error(ios,'fv_diagnostics_nml')
+        call close_file(f_unit)
+#endif
```
We are no longer supporting the monolithic open_namelist option in favor of the INTERNAL_FILE_NML logic.  Please remove the #ifdef-#endif and include only the read from input_nml_file logic.

bensonr reviewed (changes requested):
I agree with @lharris4, this is a much cleaner implementation.  Please see the review comments and address.

ywangwof commented:
Thanks for all your comments. We are still having some issues. After I have fully tested the module, I will ask for your review again.

lharris4 commented on model/fv_arrays.F90:
```diff
@@ -244,7 +244,7 @@ module fv_arrays_mod
                              !< monotonicity constraint of Huynh, which is less diffusive
                              !< but more expensive than other monotonic constraints. For hydrostatic simulation, 8
                              !< (the L04 monotonicity constraint) or 10 are recommended; for
-                             !< nonhydrostatic simulation, the completely unlimited (“linear”
```
Were these quote marks removed because of a problem with some compilers?

ywangwof replied:
Not compilers. It is my editor (vim) that cannot render those characters properly. Anyway, I have restored the file to its original format. Sorry for the trouble.

I have tested the implementation with the NSSL 2-moment MP scheme and the Thompson scheme. Unfortunately, it produces zero output for the GFDL MP scheme. John Henderson is looking at the problem for a possible reason.

lharris4 replied:
OK, thanks. 

The GFDL MP doesn't have second moments which may confuse the diagnostic.

ywangwof commented:
Hello @lharris4, @bensonr, please review this PR again and I finally had it tested with Thompson, NSSL 2-moment and GFDL microphysical schemes with the latest develop branch of the "**_ufs-weather-model_**". Thanks.

junwang-noaa commented on driver/fvGFS/fv_nggps_diag.F90:
```diff
@@ -74,6 +74,7 @@ module fv_nggps_diags_mod
  use fv_diagnostics_mod, only: range_check, dbzcalc,max_vv,get_vorticity, &
                                max_uh,max_vorticity,bunkers_vector,       &
                                helicity_relative_CAPS,max_vorticity_hy1
+ use module_diag_hailcast
```
Would you please list only those variables and subroutines in module_diag_hailcast that are used in this file?

bensonr commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1870 @@
+
+        INTEGER :: i, j
+
+        namelist /fv_diagnostics_nml/ do_hailcast
```
Although fv_diagnostics_nml is not currently being used by the FV3 fv_diagnostics.F90 module, it should be reserved for that use.  Please rename this to a more appropriate name for the hailcast module.

junwang-noaa commented on driver/fvGFS/fv_nggps_diag.F90:
```diff
@@ -770,17 +822,17 @@ subroutine fv_nggps_tavg(Atm, Time_step_atmos,avg_max_length,zvir)
-          enddo
-        enddo
+       if(mod(kdtt,nsteps_per_reset)==0)then
+            do j=jsco,jeco
```
remove extra spaces?

junwang-noaa commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1870 @@
+
+              !diag_table should select largest value of these variables during requested period
+              !thus do not need to accumulate largest value between data dumps
+              !IF (dhail1 .gt. hailcast_dhail1(i,j)) THEN
```
I find the commented if statement is confusing, do we need it or not?

junwang-noaa commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1870 @@
+        !!! Has stone reached below cloud base?
+        !IF (Z .LE. 0) GOTO 200
+        IF (Z .LE. ZBAS) then
+           GOTO 200
```
Please rewrite the code here to remove all the GOTO statement, it is not allowed in operational model code.

ywangwof commented on driver/fvGFS/fv_nggps_diag.F90:
```diff
@@ -734,6 +743,49 @@ subroutine fv_nggps_diag(Atm, zvir, Time)
       call store_data(id_uhmin25, uhmin25, Time, kstt_uhmin25, kend_uhmin25)
     endif
 
+    IF ( .not. Atm(n)%flagstruct%hydrostatic .and. do_hailcast ) THEN
```
It is impossible because it calls _`store_date`_ which is defined in **_fv_nggps_diag.F90_**. Otherwise, it will cause cyclic module dependency. 

We also considered to define a new set of subroutines of _`store_date`_, _`add_field_to_bundle`_, _`find_outputname`_ etc.  inside **_module_diag_hailcast_**. But I do not think having duplicated subroutines for the same function is a good code practice.

lharris4 replied:
I understand. This is OK, although something like `store_data` eventually needs to be defined in some better location, or adapted to use standard FMS routines.

bensonr replied:
@lharris - store_data does have take an argument that determines whether the data is handled asynchronously (buffered and sent to the write_component cores) or inline via the FMS diag_manager (send_data).

junwang-noaa commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1870 @@
+         !print *, 'INTERP VU: ', VU, P
+
+         !Outside pressure levels?  If so, exit loop
+         IF (IFOUT.EQ.1) GOTO 100
```
Please rewrite the code to remove all the GOTO statements, it is not allowed in operational model code.

junwang-noaa commented on tools/module_diag_hailcast.F90:
```diff
@@ -0,0 +1,1870 @@
+             !MEAN CLOUD DROPLET RADIUS, ASSUME CLOUD DROPLET CONC OF 3E8 M-3 (300 CM-3)
+             DC = (0.74*XW / (3.14159*1000.*3.E8))**0.33333333 * 1.E6 !MICRONS
+             !!! FIND THE STOKES NUMBER  (rasmussen heymsfield 1985)
+             NS = 2*VT*100.*(DC*1.E-4)**2. / (9*ANU*D*50)  !need hail radius in cm
```
I suggest replacing these exponent expressions with multiplications if possible, e.g. ((DC*1.E-4)**2. ->  (DC*1.E-4)*(DC*1.E-4), also W**2->W*W, W**3->W*W*W, w**4->W*W*W*W for computational efficiency.

ywangwof commented:
Thanks for your comments and suggestions. Here are the summary about  my last commit based on your comments.
1. I moved the namelist and reading from _module_diag_hailcast_ to `fv_nggps_diag_init` in _fv_nggps_diags_mod_;
2. Listed the variables and subroutines imported from _module_diag_hailcast_ explicitly in _fv_nggps_diag.F90_;
3. Removed unnecessary comments as noted by @junwang-noaa;
4. Replaced all `GOTO` statements with `EXIT` statements;
5. Introduced new variables to replaced some exponent expressions as possible.

Please review them again at your convenience. Thanks again.

laurenchilutti commented:
@ywangwof is there a fv3atm and ufs-weather-model PR in place to bring these changes into the UFS weather model?

ywangwof commented:
Thanks. @laurenchilutti. I just issued two PRs for them.
fv3atm, NOAA-EMC/fv3atm/pull/518.
ufs-weather-model, ufs-community/ufs-weather-model/pull/1164

junwang-noaa commented:
@laurenchilutti @bensonr FYI. we are running RT in UFS (PR#1156)[https://github.com/ufs-community/ufs-weather-model/pull/1156] with this PR. We will let you know when all the RT test pass, please commit the changes (probably on 4/29). Thanks

binli2337 commented:
@laurenchilutti @bensonr  All tests passed in UFS (PR#1156). Please commit the changes from PR#164. Thanks.
